src/NuGet.Core/NuGet.Protocol/Resources/PackageMetadataResourceV3.cs

```diff
@@ -252,7 +254,7 @@ private async Task<T> DeserializeStreamDataAsync<T>(Stream stream, CancellationT
         /// <param name="includeUnlisted">Whether to include Unlisted versions into result.</param>
         private void ProcessRegistrationPage(
             RegistrationPage registrationPage,
-            List<PackageSearchMetadataRegistration> results,
+            ConcurrentBag<PackageSearchMetadataRegistration> results,
```

Comment:
I think you are likely better off by using a lock instead of a concurrent bag. 
A concurrent bag is unordered, and these lists will usually be sorted later, so sorts are significantly this way.

Furthermore, `lock` will be a lot faster. 

Beyond that, it seems reasonable, since this affects the processing of pages only.

Reply:
I dont think a lock would be better if your concern is an ordered list as there is no guarantee of the order of the files being processed finishing in any specific order.  If the results need to be sorted I would prefer to just use a sorting method before returning.

I wasn't sure if the ordering really mattered, the main target that I know of using this data does its own sort for instance.

Reply:
I too think that we should preserve the existing order.

Reply:
If you pre-allocate the list (add `null` values, before starting the work in parallel), then if each task knows which index is theirs, this can work lock-free. If any of the tasks fail, you need to be able to handle a null entry in the list, but from the look of the changes, in both the before and after case, we should be propagating the exception, so the list doesn't even get returned.

Reply:
Alternatively, if the `async delegate` returns a `Task<PackageSearchMetadataRegistration>`, then you can have

```cs
List<Task<PackageSearchMetadataRegistration>> tasks  = registrationIndex.Items.Select(GetItemAsync);
await Task.WhenAll(tasks);
List<PackageSearchMetadataRegistration> = tasks.Select(task => task.Result);
```

Again, lock free.

Reply:
@zivkan Thats a great idea, thank you for the suggestion.

Reply:
After playing around with it I decided that a simple sort after the fact was cleanest.

Reply:
Does perf gain still hold after switching to `ConcurrentBag` and `ConcurrentDictionary`?